docs/design_proposals/configurable-kubecontext.md

```diff
@@ -0,0 +1,201 @@
+
+For example, the following sequence will result in an error:
+
+- The current context is `minikube` and activates some profile.
+- Some profile also overrides the kubecontext and deploys to kubecontext `gke_abc_123`.
+- Thus the `minikube`-specific profile would be deployed to `gke_abc_123`, and this will be forbidden.
+
+**Note**: `skaffold.yaml` is meant to be shared, but kubecontext names vary across users.
+Sharing therefore makes only sense in a corporate setting where context names are the same across different machines.
+
+#### Option in global Skaffold config
+Specifying a default kubecontext globally is straightforward. For example, via new config option
+```yaml
+global:
+  default-context: docker-for-desktop
+```
+
+On the other hand, building a relation between projects and kubecontext needs to solve two questions:
+
+1. How to identify projects
+2. How to save the project/kubecontext relation
+
+##### How to identify projects
+
+There are at least three possibilities:
+
+- Identify projects by their absolute host path.
+  This is guaranteed to be unique, but may break if a user moves his project to another location.
+- Identify projects by a new `metadata.name` entry in `skaffold.yaml` (see also [#2200](https://github.com/GoogleContainerTools/skaffold/issues/2200)).
+  This has the drawback of being potentially not unique, so that users accidentally pin the kubecontext for more projects than intended.
+  On the other hand, this is the standard approach taken by kubernetes resources.
+- Identify project by their initial commit.
+  This variant is stable against relocations.
+  It is also unique unless a user forks a project and wants to define different kubecontexts for each fork.
+  Compared to the other alternatives, it is rather slow.
+
+**\<What option has the best tradeoffs?\>**
+
+Resolution: We will go ahead with the `metadata.name` approach. As the name may not be unique, this requires special documentation.
+
+
+##### How to save project/kubecontext relations
+
+Currently, the Skaffold config uses the kubecontext as identifier.
+
+There are two possibilities to add the relation:
+
+- Reverse the mapping project/kubecontext and save as list under `kube-context` entries:
+  ```yaml
+  kubecontexts:
+  - kube-context: my-context
+    skaffoldConfigs:
+    - config-name
+  ```
+  The drawback here is that the data structure does not forbid conflicting entries, such as this:
+  ```yaml
+  kubecontexts:
+  - kube-context: context1
+    skaffoldConfigs:
+    - my-project
+  - kube-context: context2
+    skaffoldConfigs:
+    - my-project
+  ```
+- Add a new top-level entry in Skaffold config:
+  ```yaml
+  global: {}
+  kubecontexts: []
+  projects:
+    my-project: # project name as the key
+      kube-context: context1
+  ```
+  This option will be more complex to implement wrt `skaffold config`.
+
+### Open Issues/Questions
+
+**\<What Skaffold config structure has the best tradeoffs?\>**
+
+Resolution: __Not Yet Resolved__
```

Comment:
@balopat Let's resolve this last open question, so that this design proposal can be merged.

My feeling is that option 2 (map-like structure) is overall better suited to store project-to-context mappings. Besides, option 1 (`skaffoldConfigs` list under kubecontext entry) requires additional special cases for the `skaffold config set/unset` commands. So I'm more inclined towards the second option:
```yaml
skaffoldConfigs:
  my-project: my-context
  my-other-project: my-other-context
```
WDYT?

Reply:
I just found another advantage of option 2: It allows a much more seamless implementation of the wildcard global default context like so:
```yaml
skaffoldConfigs:
  '*': default-context
  my-project: my-context
```

Reply:
I'm okay with this direction!